### PR TITLE
Renovated Mutations RegEx, refactored MutText normalization, added tests for Mutation.extractGene

### PR DESCRIPTION
**Pull request by arodcasanova**

### CODE REVIEW

File: Utilities/src/test/java/edu/stanford/hivdb/mutations/MutationTest.java [hunk 1]
```diff
@@ -428,6 +436,7 @@ public void testGetHumanFormat() {
 
 	@Test
 	public void testParseString() {
+		Mutation.parseString(Gene.RT, "T69Insertion");
```
Comment (philiptzou) [inline Utilities/src/test/java/edu/stanford/hivdb/mutations/MutationTest.java]: No assertion?

File: Utilities/src/test/java/edu/stanford/hivdb/mutations/MutationTest.java [hunk 2]
```diff
@@ -85,13 +85,21 @@ public void testNormalizeAAs() {
...
+	public void testExtractGene() {
+		assertEquals(Gene.PR, Mutation.extractGene("PR100A"));
+		assertEquals(Gene.IN, Mutation.extractGene("IN100A"));
+		assertEquals(Gene.RT, Mutation.extractGene("RT100A"));
+		assertEquals(null, Mutation.extractGene("not a mutation"));
+	}
+	
+	@Test(expected=InvalidMutationStringException.class)
+	public void testExtractGeneWithMalformedMut() {
+		assertEquals(null, Mutation.extractGene("P100D"));
```
Comment (philiptzou) [inline Utilities/src/test/java/edu/stanford/hivdb/mutations/MutationTest.java]: Can you explain why `P100D` invalid again? I'm confused.

File: Utilities/src/test/java/edu/stanford/hivdb/mutations/MutationTest.java (hunk 1 above)
Comment (arodcasanova) [reply Utilities/src/test/java/edu/stanford/hivdb/mutations/MutationTest.java]: I actually meant to remove this line, but I think it's a good test case to include. So I've asserted it.

File: Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java [hunk 3]
```diff
@@ -135,28 +136,19 @@ public static Mutation fromNucAminoMutation(Gene gene, int aaStart, Map<?, ?> mu
...
-		else if (aas.equals("Deletion")) {
-			aas = "-";
-		}
-		if (aas.length() > 1 &&
-			   	!aas.contains("_")) {
-			// sort mixture
-			aas = MyStringUtils.sortAlphabetically(aas).toUpperCase();
+		if (aas == null) return null;
+		
+		aas = aas.replaceAll("[dD]eletion|d(el)?|~", "-")
```
[the remark this answers is not in the document]
Comment (arodcasanova) [reply Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java]: Should mutationPattern also accept "delete"?

File: Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java [hunk 4]
```diff
@@ -135,28 +136,19 @@ public static Mutation fromNucAminoMutation(Gene gene, int aaStart, Map<?, ?> mu
...
+			     .replaceAll("[iI]nsertion|i(ns)?|#", "_")
```
[the remark this answers is not in the document]
Comment (arodcasanova) [reply Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java]: Should mutationPattern also accept "insert"?

File: Utilities/src/test/java/edu/stanford/hivdb/mutations/MutationTest.java (hunk 2 above)
Comment (arodcasanova) [reply Utilities/src/test/java/edu/stanford/hivdb/mutations/MutationTest.java]: While P100D is a valid mutation string, it is invalid in the context of `extractGene()`. Specifically, `extractGene()`, chooses to throw an `InvalidMutationStringException` if `Gene.valueOf(m.group(1).toUpperCase())` fails. 

Indeed, throwing this specific exception is a bit misleading.

File: Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (philiptzou) [reply Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java]: Well, it's kinda debatable. However, we didn't accept the verbs "delete" and "insert" in the past and it didn't cause any problem. We'd better just keep it as simple as it was.

File: Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (philiptzou) [reply Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java]: See my response above.

File: Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (arodcasanova) [reply Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java]: Gotcha. Did you mean to type "delete"? I'm a bit confused about which test to add. Should I add a test to assert "delete" isn't captured?

[the remark this answers is not in the document]
Comment (philiptzou) [reply Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java]: I mean, when someone typed "delet", it shouldn't be replaced to "-et". Following are some examples of allows and disallows.

Allows:

T69Insertion, T69Deletion, T69insertion, T69deletion, T69ins, T69del, T69i, T69d, T69#, T69~, T69T#D, T69_, T69T_D, T69INS (three amino acids, not insertion), T69DEL (three amino acids, not deletion)

Disallows:

T69[Ii]nser, T69[Dd]elet, T69~ACD, T69#ACD, T69T#, T69T~, T69TinsD, T69TdelD...

[the remark this answers is not in the document]
Comment (arodcasanova) [reply Utilities/src/main/java/edu/stanford/hivdb/mutations/Mutation.java]: Awesome. Thanks for that clarification.

File: Utilities/src/test/java/edu/stanford/hivdb/mutations/MutationTest.java (hunk 2 above)
Comment (arodcasanova) [reply Utilities/src/test/java/edu/stanford/hivdb/mutations/MutationTest.java]: Oo good tip.

**Comment by philiptzou** [comment]
@arodcasanova Thank you!

**Pull request merged**